### ISSUE TITLE
Reducing startup time

### ISSUE THREAD

**Issue by DanOscarsson** [libass/libass#564]
I have been looking at startup time for a while. I have mpv which uses libass for subs and on screen info. mpv creates a libass instance (library object, render, track, etc) on the fly. Often at least 3 instances. And when a subtitle track is switched it will create a new instance for the new track and destroy the old one.
On my Linux system creating one instance takes about 10-20 ms.
This is big enough to sometimes disrupt the normal flow of video frames.
One way to reduce this is create an object shared by all instances where sharable data could be stored.
On my system most of the startup time is in fontconfig when it reads the system configuration data and fonts. So I have tested to place the fontconfig instance in a sharable object.
This results in first use of 10-20 ms startup, like before. But for all new instances after that I get about 2 ms.
I suspect other parts of font data, like caches, could also be put in the shared object.

That is my background. Now I wonder what your thoughts are about this?
My idea is to start by creating a sharable object, with some suitable name, that can be initialised before ASS_library and then given to all ASS_library instances where you want data to be shared. Then hopefully heavy initialisation can be done with this object so it only have to be done once.
While my tests show good results it will take time before I can come back with a ready PR.
But it would be nice to design it in a way you prefer, if you are interested?

**Comment by astiob** [comment]
That’s what `ASS_Library` itself is. There is a known issue that track-specific fonts get added to the library object, but I’m not sure that’s the original reason why mpv isn’t sharing its `ASS_Library`. I seem to recall something else… maybe multithreading?

**Comment by DanOscarsson** [comment]
Yes, the main problem I see with ASS_Library are fonts. From mpv and the code there are, apart from system fonts, memory based and embedded. mpv uses momory fonts in two contexts: OSD and media file. For OSD it adds a symbol font, and for media files it adds any attached fonts.
Then, from what I can see from the code, ass can have embedded fonts per track.
So there should be a "context" object that have the memory fonts and the track object should contain the embedded fonts. Then it would be much easier.
Looking at older issues I see that the different objects have been discussed now and then.
From what I can see in the code the ASS_Renderer is intended to be used with one ASS_Track as there are many things in it related to current track and the events in it that is being rendered.
One way that might enable ASS_Renderer to be used for many tracks and as "context" object would be to move track specific attributes to the ASS_Track object (or a render object attached to the track object).
Maybe much of changes can be hidden so current API can mostly be unchanged.

**Comment by astiob** [comment]
It’s been pointed out that each `ASS_Renderer` can have its own font configuration (particularly its own Fontconfig configuration in your case) via `ass_set_fonts`, so any change here would probably require an API break.

**Comment by DanOscarsson** [comment]
One problem I can see with using the same ASS_Renderer in parallel with events from different tracks are that images_root and prev_images_root are not in the track object. A test for changed render will result in changes all the time as you will compare rendering of events from different tracks.

**Comment by rcombs** [comment]
Can we examine why fontconfig is taking so long to initialize? Is this necessary?

Also, mpv could probably stop replacing the library and renderer when switching subtitle streams, and instead only replace the ASS_Track (with a library+renderer instead being created whenever loading a new file).

**Comment by DanOscarsson** [comment]
Now I have tested some ideas and as renderer and track can be created and destroyed in any order (for example mpv can destroy renderer before the tracks it has rendered) some things are not easy to change. I tried to move renderer images_root to track but that will not work if renderer is destroyed before track as they point to cached objects in renderer.
So change detection will only work well as long as you render events of same track in sequence instead of switching back and forth between tracks in one renderer. Probably what most do.

Then I looked at a way to avoid having memory fonts in library object. Looked messy but now I think I have a good solution. I have introduced a "font container" object.
By default the library object have a container and things work like today.
But if you want you create a "font container" object, add your own memory fonts to it and then attach it to track object and render object. If that is don, embedded fonts parsed by track goes into "font container" and renderer loads memory fonts from it instead of from library object.
This way you can have one library object for all tracks and renderers like today, with shared memory fonts. And you can, by creating "font container" object, still share one library object while letting different tracks and renderers have different memory fonts.
Does that sound like a solution ok for you?
